### PR TITLE
Some preparatory work for DTLS client hello defragmentation

### PR DESCRIPTION
**Pull request by ronald-cron-arm**
## Description
Create branch specific generate_tls_handshake_test.py and make_generated_files.py.
generate_tls_handshake_test.py will be used to specifiy that the branch supports (D)TLS 1.2 ClientHello defragmentation.
make_generated_files.py will be used to specify that the branch specific generate_tls_handshake_test.py should be used instead of the temporary framework one.

Depends on: https://github.com/Mbed-TLS/mbedtls-framework/pull/274

## PR checklist
- [x] **changelog** not required because: no change from the user perpspective
- [x] **development PR** here
- [x] **TF-PSA-Crypto PR** not required because: no need for any change in TF-PSA-Crypto
- [x] **3.6 PR** provided #10603
- **tests**  not required because: changing the organization of utility scripts and TLS logs improvements

### CODE REVIEW

**Comment by ronald-cron-arm** [comment]
@minosgalanakis @bensze01 as mentioned offline I've updated the PR with new commits. CI is happy. The PR is ready for review, please have a look.

File: scripts/generate_tls_handshake_tests.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+# SPDX-License-Identifier: Apache-2.0 OR GPL-2.0-or-later
+
+import sys
+
+import framework_scripts_path # pylint: disable=unused-import
+
+from mbedtls_framework import tls_handshake_tests
+
+if __name__ == '__main__':
+    sys.argv[1:1] = ["--no-tls12-client-hello-defragmentation-support"]
```
Comment (minosgalanakis) [inline scripts/generate_tls_handshake_tests.py:16]: Small note. This file should be in the state that the linked  pr [is approved at.](https://github.com/Mbed-TLS/mbedtls-framework/pull/274#discussion_r2763784386)

File: tests/ssl-opt.sh [hunk 2]
```diff
@@ -11496,9 +11496,9 @@ run_test    "DTLS reordering: Buffer out-of-order handshake message on client" \
             hs_timeout=2500-60000" \
             0 \
             -c "Buffering HS message" \
-            -c "Next handshake message has been buffered - load"\
+            -c "Certificate handshake message has been buffered$"\
```
Comment (minosgalanakis) [inline tests/ssl-opt.sh:11504]:
```suggestion
            -c "Certificate handshake message has been buffered"\
```

File: tests/ssl-opt.sh [hunk 3]
```diff
@@ -11539,10 +11539,11 @@ run_test    "DTLS reordering: Buffer out-of-order hs msg before reassembling nex
             hs_timeout=2500-60000" \
             0 \
             -c "Buffering HS message" \
-            -c "Next handshake message has been buffered - load"\
+            -c "Certificate handshake message has been buffered and reassembled"\
+            -c "ServerKeyExchange handshake message has been buffered$"\
```
Comment (minosgalanakis) [inline tests/ssl-opt.sh:11548]:
```suggestion
            -c "ServerKeyExchange handshake message has been buffered"\
```

File: tests/ssl-opt.sh [hunk 4]
```diff
@@ -11582,9 +11583,9 @@ run_test    "DTLS reordering: Buffer out-of-order handshake message on server" \
             hs_timeout=2500-60000" \
             0 \
             -C "Buffering HS message" \
-            -C "Next handshake message has been buffered - load"\
+            -C "handshake message has been buffered" \
             -s "Buffering HS message" \
-            -s "Next handshake message has been buffered - load" \
+            -s "ClientKeyExchange handshake message has been buffered$" \
```
Comment (minosgalanakis) [inline tests/ssl-opt.sh:11593]:
```suggestion
            -c "ClientKeyExchange handshake message has been buffered"\
```

File: tests/ssl-opt.sh [hunk 5]
```diff
@@ -11863,21 +11863,23 @@ not_with_valgrind # risk of non-mbedtls peer timing out
 requires_config_enabled MBEDTLS_SSL_PROTO_TLS1_2
 run_test    "DTLS proxy: 3d, openssl server, fragmentation" \
             -p "$P_PXY drop=5 delay=5 duplicate=5 protect_hvr=1" \
-            "$O_NEXT_SRV -dtls1_2 -mtu 768" \
-            "$P_CLI dgram_packing=0 dtls=1 hs_timeout=500-60000 tickets=0" \
+            "$O_NEXT_SRV -dtls1_2 -mtu 256" \
```
Comment (minosgalanakis) [inline tests/ssl-opt.sh:11866]: I understand that we need to reduce that mtu to force the handshake to be fragmented, but I am not sure why it was set to 768 before. Do you think that was arbitrary?

Comment (ronald-cron-arm) [reply tests/ssl-opt.sh:11866]: I can see that the record containing the Certificate message received by the client is 579 bytes long. Thus with an MTU  of 768 there is no fragmentation. The server certificate is server5.crt. The PEM files is 798 bytes long thus maybe the MTU was set below this value. But in the certificate message the encoding is more compact.

**Comment by ronald-cron-arm** [comment]
@minosgalanakis @bensze01 I've added three commits. One to update the framework pointer to the merge of f#274 and two to fix two issues I spotted when working on the 3.6 backport. CI green, still ready for review.

File: tests/compat.sh [hunk 6]
```diff
@@ -602,7 +604,14 @@ setup_arguments()
...
         O_CLIENT_ARGS="$O_CLIENT_ARGS -CAfile $DATA_FILES_PATH/test-ca_cat12.crt -verify 10"
-        G_CLIENT_ARGS="$G_CLIENT_ARGS --x509cafile $DATA_FILES_PATH/test-ca_cat12.crt"
+
+        # Use --insecure for the GnuTLS client because it expects the
+        # hostname/IP it connects to to be the name used in the certificate
+        # obtained from the server. Here, however, it connects to 127.0.0.1
+        # while our test certificates use 'localhost' as the server name in the
+        # certificate. This will make the certificate validation fail, but
+        # passing --insecure makes GnuTLS continue the connection nonetheless.
+        G_CLIENT_ARGS="$G_CLIENT_ARGS --insecure --x509cafile $DATA_FILES_PATH/test-ca_cat12.crt"
```
Comment (bensze01) [inline tests/compat.sh]: How about this?
```suggestion
        G_CLIENT_ARGS="$G_CLIENT_ARGS --verify-hostname=localhost --x509cafile $DATA_FILES_PATH/test-ca_cat12.crt"
```

File: scripts/generate_tls_handshake_tests.py (hunk 1 above)
Comment (minosgalanakis) [reply scripts/generate_tls_handshake_tests.py:16]: This has been merged as such so this issue is resolved.

File: tests/compat.sh (hunk 6 above)
Comment (ronald-cron-arm) [reply tests/compat.sh]: gnutls-cli 3.4.6 does not support the option "--verify-hostname" unfortunately. Thinking about that I should not introduce "--insecure" here. That's dangerous. We could miss things in the future because of that. I am reverting the change and adapt the comment to help if somebody hits the same issue locally.

File: library/ssl_tls.c [hunk 7]
```diff
@@ -694,8 +694,16 @@ static const char *ssl_tls13_get_hs_msg_name(int hs_msg_type)
             return "EncryptedExtensions";
         case MBEDTLS_SSL_HS_CERTIFICATE:
             return "Certificate";
+        case MBEDTLS_SSL_HS_SERVER_KEY_EXCHANGE:
```
Comment (minosgalanakis) [inline library/ssl_tls.c:697]: This should match the method on [this](https://github.com/Mbed-TLS/mbedtls/pull/10576#discussion_r2817709446) pr

**Review by minosgalanakis — COMMENTED** [review]
Looks good overall. I am happy to approve at the current state, if the [comment](https://github.com/Mbed-TLS/mbedtls/pull/10575#discussion_r2817252920) proves to have too many edge cases to implement.

File: tests/compat.sh (hunk 6 above)
Comment (ronald-cron-arm) [reply tests/compat.sh]: Eventually I've just removed completely the commit "compat.sh: Amend GnuTLS client arguments". It was just not doing the right things.

**Comment by ronald-cron-arm** [comment]
@minosgalanakis @bensze01 I have rebased due to a framework conflict. There is only one "update framework pointer" commit now, the first one. Otherwise, as mentioned [here](https://github.com/Mbed-TLS/mbedtls/pull/10575#discussion_r2817252920), I've removed the commit "compat.sh: Amend GnuTLS client arguments". CI is green, please have another look.

**Review by bensze01 — APPROVED** [review]
Thanks, LGTM!
